summingbird-online/src/main/scala/com/twitter/summingbird/planner/OnlinePlan.scala

```diff
@@ -143,10 +156,10 @@ class OnlinePlan[P <: Platform[P], V](tail: Producer[P, V]) {
           case _ if (!noOpNode(activeBolt) && dependsOnSummerProducer(currentProducer)) => true
           /*
            * This should possibly be improved, but currently, we force a FlatMapNode just before a
-           * summer (to handle map-side aggregation). This check is here to prevent us from merging
-           * this current node all the way up to the source.
+           * summer (to handle map-side aggregation) unless the currentProducer is configured to be merged into Source.
+           * This check is here to prevent us from merging this current node all the way up to the source.
            */
-          case FlatMapNode(_) if hasSummerAsDependantProducer(currentProducer) && allTransDepsMergeableWithSource(dep) => true
+          case FlatMapNode(_) if hasSummerAsDependantProducer(currentProducer) && allTransDepsMergeableWithSource(dep) && !(mergableWithSource(currentProducer)) => true
           /*
```

Comment:
The condition is becoming too complex for a single line, may be refactor to a separate function.